pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go

```diff
@@ -0,0 +1,241 @@
+	}
+
+	// Have the informer bust cache entries when it sees updates that could
+	// apply to them.
+	_, err := m.ctbInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			ctb, ok := obj.(*certificatesv1alpha1.ClusterTrustBundle)
+			if !ok {
+				return
+			}
+			klog.InfoS("Dropping all cache entries for signer", "signerName", ctb.Spec.SignerName)
+			m.dropCacheFor(ctb)
+		},
+		UpdateFunc: func(old, new any) {
+			ctb, ok := new.(*certificatesv1alpha1.ClusterTrustBundle)
+			if !ok {
+				return
+			}
+			klog.InfoS("Dropping cache for ClusterTrustBundle", "signerName", ctb.Spec.SignerName)
+			m.dropCacheFor(new.(*certificatesv1alpha1.ClusterTrustBundle))
+		},
+		DeleteFunc: func(obj any) {
+			ctb, ok := obj.(*certificatesv1alpha1.ClusterTrustBundle)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					return
+				}
+				ctb, ok = tombstone.Obj.(*certificatesv1alpha1.ClusterTrustBundle)
+				if !ok {
+					return
+				}
+			}
+			klog.InfoS("Dropping cache for ClusterTrustBundle", "signerName", ctb.Spec.SignerName)
+			m.dropCacheFor(ctb)
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("while registering event handler on informer: %w", err)
+	}
+
+	return m, nil
+}
+
+func (m *InformerManager) dropCacheFor(ctb *certificatesv1alpha1.ClusterTrustBundle) {
+	if ctb.Spec.SignerName != "" {
+		m.normalizationCache.RemoveAll(func(key any) bool {
+			return key.(cacheKeyType).signerName == ctb.Spec.SignerName
+		})
+	} else {
+		m.normalizationCache.RemoveAll(func(key any) bool {
+			return key.(cacheKeyType).ctbName == ctb.ObjectMeta.Name
+		})
+	}
+}
+
+func (m *InformerManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]byte, error) {
+	if !m.ctbInformer.HasSynced() {
+		return nil, fmt.Errorf("ClusterTrustBundle informer has not yet synced")
+	}
+
+	ctb, err := m.ctbLister.Get(name)
+	if k8serrors.IsNotFound(err) && allowMissing {
+		return []byte{}, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("while getting ClusterTrustBundle: %w", err)
+	}
+
+	cacheKey := cacheKeyType{ctbName: name}
+
+	if cachedAnchors, ok := m.normalizationCache.Get(cacheKey); ok {
+		return cachedAnchors.([]byte), nil
+	}
+
+	pemTrustAnchors, err := m.normalizeTrustAnchors([]*certificatesv1alpha1.ClusterTrustBundle{ctb})
+	if err != nil {
+		return nil, fmt.Errorf("while normalizing trust anchors: %w", err)
+	}
+
+	m.normalizationCache.Add(cacheKey, pemTrustAnchors, 5*time.Minute)
+
+	return pemTrustAnchors, nil
+}
+
+func (m *InformerManager) GetTrustAnchorsBySigner(signerName string, labelSelector *metav1.LabelSelector, allowMissing bool) ([]byte, error) {
+	if !m.ctbInformer.HasSynced() {
+		return nil, fmt.Errorf("ClusterTrustBundle informer has not yet synced")
+	}
+
+	// Note that this function treats nil as "match nothing", and non-nil but
+	// empty as "match everything".
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
```

Comment:
Do we care if someone tries to DOS the cache with a large label selector?

Reply:
Added a 100KiB limit.  I am not sure that there is an additional concern here, since such a large label selector would also be DoSing kubelet in other ways (the Pod object would be very large, for example).